BUG: fix ttest zero division handling

andim opened the pull request:
The ttest function in `scipy.stats` currently do not check for zero division caused by variances being zero. The behavior of a function call thus depends on the numpy error state. Depending on this state their might either be annoying warnings or exceptions will be raised.

This PR fixes these issues by performing explicit checks before division.

andim commented:
Up to now the functions returned `np.nan` for both statistic and pvalue if both the variance and the difference of the means is zero (if numpy errors where ignored). It seems more consistent to me to return a statistic of 0.0 and a pvalue of 1.0. I've changed the code and test case to this effect. If there are valid objections to why the old (undocumented) behavior should remain I am happy to revert this part of the fix.

josef-pkt commented on scipy/stats/tests/test_stats.py:
```diff
@@ -2451,12 +2451,12 @@ def test_ttest_ind():
+    # test zero division problem
+    t,p = stats.ttest_ind([0,0,0], [1,1,1])
+    assert_equal((np.abs(t),p), (np.inf, 0))
+    assert_equal(stats.ttest_ind([0,0,0], [0,0,0]), (0.0, 1.0))
```
We had the discussion about the zero variance case before and decided that it is supposed to return nan.
IIRC, this is the case where we changed it.

andim replied:
It is supposed to return nan no matter the denominator?

josef-pkt replied:
I'm not sure I remember the details correctly. 
Based on the above unit tests, the discussion was just for the 0/0 = nan case.  2/0 is well defined as inf.

andim replied:
I've found part of the discussion by searching through the scipy-dev mailing list : http://article.gmane.org/gmane.comp.python.scientific.devel/16645/

Sorry, I have not looked there before. Haven't been around for so long ;)

josef-pkt commented:
LGTM, reading through the changes. 
One of the test run stops on a pep-8 failure. And I'm not familiar enough anymore with the code to guess if there are no problems that are not visible in the change set.

andim commented:
Test no longer failing after latest fix. Any other changes needed?

ev-br commented on scipy/stats/stats.py:
```diff
@@ -3707,7 +3709,8 @@ def _ttest_ind_from_stats(mean1, mean2, denom, df):
     vn2 = v2 / n2
-    df = ((vn1 + vn2)**2) / ((vn1**2) / (n1 - 1) + (vn2**2) / (n2 - 1))
+    with np.errstate(divide='ignore', invalid='ignore'):
+        df = ((vn1 + vn2)**2) / ((vn1**2) / (n1 - 1) + (vn2**2) / (n2 - 1))
```
Optional: while you're touching this line, it'd be nice to remove extra brackets: 
```
(vn1 + vn2)**2 / (vn1**2 / (n1 - 1) + vn2**2 / (n2 - 1))
```

looks a bit more readable.

andim replied:
done.

ev-br commented on scipy/stats/tests/test_stats.py:
```diff
@@ -2358,13 +2358,13 @@ def test_ttest_rel():
     assert_raises(ValueError, stats.ttest_rel, x, y, nan_policy='foobar')
 
+    # test zero division problem
+    t,p = stats.ttest_rel([0,0,0],[1,1,1])
```
Also optional: spaces after commas please. There's no need to go haywire correcting PEP8 violations in random places, but once you've touched a line, it's nice to also improve readability.

andim replied:
done.

ev-br commented:
One corner case is if the input contains `nan`s and `nan_policy='omit'`:  

```
In [11]: stats.ttest_rel([0,0, np.nan],[1,1,1])
Out[11]: Ttest_relResult(statistic=array(nan), pvalue=nan)       # OK

In [12]: stats.ttest_rel([0,0, np.nan],[1,1,1], nan_policy='omit')
/home/br/installs/scipy/build/testenv/lib/python2.7/site-packages/scipy/stats/mstats_basic.py:953: RuntimeWarning: divide by zero encountered in double_scalars
  t = ma.add.reduce(d, axis) / denom
```

This is because `nan_policy='omit'` delegates to the mstats version. Would you please make a similar change to `mstats.ttest_*`s

ev-br commented:
In addition to what Josef said, I made two trivial nitpicks and one easy-to-implement comment. 
And the last nitpick --- would be good to squash the commits if you are at ease with `git rebase -i`. If not, don't worry about it.

Other than that, LGTM.

andim commented:
I've also added the changes to `mstats.ttest_*` along with porting the zero division unittests from the non-masked case. This is when I noticed that the behavior of the `mstats` function so far differed in its return for the 0/0 case. In fixing this issue I fixed an unrelated bug which lead to `mstats.ttest_*` returning non-sensical values if one of the arrays is fully masked. Then t is calculated to be `np.nan`, which was then replaced by 1. So you got a statistic of 1 and a finite pvalue depending on the size of the other array. This should now hopefully be fixed (returning `np.nan` instead) and I added test cases to prevent regressions.

ev-br commented on scipy/stats/mstats_basic.py:
```diff
@@ -839,10 +839,11 @@ def ttest_1samp(a, popmean, axis=0):
+    with np.errstate(divide='ignore', invalid='ignore'):
+        svar = ((n - 1.0) * v) / df
+        t = (x - popmean) / ma.sqrt(svar / n)
+        prob = special.betainc(0.5*df, 0.5, df/(df + t*t))
```
I don't quite understand why one needs to move *more* into the `with np.errstate` block. In the `stats` version, it's just one single division, why cannot we do the same here?
(manipulating the errstate might mask other bugs?)

andim replied:
I get a number of warnings of the following type for the fully masked test otherwise. 
`RuntimeWarning: divide by zero encountered in true_divide
probs = special.betainc(0.5*df, 0.5, df/(df + t*t)).reshape(t.shape)`

I agree it's not an elegant solution. If you have a better idea let me know.

andim replied:
I tracked down the source of the error a bit further. It boils down to the following throwing an Runtime warning:
np.float64(1.0) / ma.masked_array([0.0], mask=[1.0])
Surprisingly (?) the following does not have the same problem
`print(type(np.array([1.0]) - 1.0))`

andim commented:
I've added explicit conversions (`df= ma.asanyarray(n-1.0)`) to ensure proper masked division when the array is fully masked. I've also changed the `ttest_rel` code to avoid having same issue in the calculation of the variance.

ev-br commented on scipy/stats/tests/test_mstats_basic.py:
```diff
@@ -754,6 +760,10 @@ def test_empty(self):
+    def test_zero_division(self):
+        t, p = mstats.ttest_rel([0, 0, 0], [1, 1, 1])
+        assert_equal((np.abs(t), p), (np.inf, 0))
+        assert_array_equal(mstats.ttest_ind([0, 0, 0], [0, 0, 0]), (np.nan, np.nan))
```
`ttest_rel`?

andim commented:
Should I squash the commits down to one? It seems like the PR is now a quite self-contained solution to the issue at hand. There is definitely further unrelated work needed on the `mstats.ttest_*` code, but that's maybe best for a separate pull request.

ev-br commented:
Yes please. One or two or three but definitely not thirteen :-).

andim commented:
Squashed down to two commits. One for the stats fixes and another one for the mstats fixes. Thanks for your comments and looking forward to seeing this fixed.

ev-br commented:
Thank you @andim, keep them coming!
